src/provider/hyrax_pc.rs

```diff
@@ -0,0 +1,414 @@
+    let result = (self as &HyraxCommitment<G>)
+      .comm
+      .iter()
+      .map(|c| c * &scalar)
+      .collect();
+    *self = HyraxCommitment {
+      comm: result,
+      is_default: self.is_default,
+    };
+  }
+}
+
+impl<'a, 'b, G: Group> Mul<&'b G::Scalar> for &'a HyraxCommitment<G> {
+  type Output = HyraxCommitment<G>;
+  fn mul(self, scalar: &'b G::Scalar) -> HyraxCommitment<G> {
+    let result = self.comm.iter().map(|c| c * scalar).collect();
+    HyraxCommitment {
+      comm: result,
+      is_default: self.is_default,
+    }
+  }
+}
+
+impl<G: Group> Mul<G::Scalar> for HyraxCommitment<G> {
+  type Output = HyraxCommitment<G>;
+
+  fn mul(self, scalar: G::Scalar) -> HyraxCommitment<G> {
+    let result = self.comm.iter().map(|c| c * &scalar).collect();
+    HyraxCommitment {
+      comm: result,
+      is_default: self.is_default,
+    }
+  }
+}
+
+impl<'b, G: Group> AddAssign<&'b HyraxCommitment<G>> for HyraxCommitment<G> {
+  fn add_assign(&mut self, other: &'b HyraxCommitment<G>) {
+    if self.is_default {
+      *self = other.clone();
+    } else if other.is_default {
+      return;
+    } else {
+      let result = (self as &HyraxCommitment<G>)
+        .comm
+        .iter()
+        .zip_longest(other.comm.iter())
+        .map(|x| match x {
+          Both(a, b) => a + b,
+          Left(a) => *a,
+          Right(b) => *b,
+        })
+        .collect();
+      *self = HyraxCommitment {
+        comm: result,
+        is_default: self.is_default,
+      };
+    }
+  }
+}
+
+impl<'a, 'b, G: Group> Add<&'b HyraxCommitment<G>> for &'a HyraxCommitment<G> {
+  type Output = HyraxCommitment<G>;
+  fn add(self, other: &'b HyraxCommitment<G>) -> HyraxCommitment<G> {
+    if self.is_default {
+      other.clone()
+    } else if other.is_default {
+      self.clone()
+    } else {
+      let result = self
+        .comm
+        .iter()
+        .zip_longest(other.comm.iter())
+        .map(|x| match x {
+          Both(a, b) => a + b,
+          Left(a) => *a,
+          Right(b) => *b,
+        })
+        .collect();
+      HyraxCommitment {
+        comm: result,
+        is_default: self.is_default,
+      }
+    }
+  }
+}
+
+macro_rules! define_add_variants {
+  (G = $g:path, LHS = $lhs:ty, RHS = $rhs:ty, Output = $out:ty) => {
+    impl<'b, G: $g> Add<&'b $rhs> for $lhs {
+      type Output = $out;
+      fn add(self, rhs: &'b $rhs) -> $out {
+        &self + rhs
+      }
+    }
+
+    impl<'a, G: $g> Add<$rhs> for &'a $lhs {
+      type Output = $out;
+      fn add(self, rhs: $rhs) -> $out {
+        self + &rhs
+      }
+    }
+
+    impl<G: $g> Add<$rhs> for $lhs {
+      type Output = $out;
+      fn add(self, rhs: $rhs) -> $out {
+        &self + &rhs
+      }
+    }
+  };
+}
+
+macro_rules! define_add_assign_variants {
+  (G = $g:path, LHS = $lhs:ty, RHS = $rhs:ty) => {
+    impl<G: $g> AddAssign<$rhs> for $lhs {
+      fn add_assign(&mut self, rhs: $rhs) {
+        *self += &rhs;
+      }
+    }
+  };
+}
```

Comment:
Consider using [`forward_ref`](https://crates.io/crates/forward_ref) or even, if need be, [`forward_ref_generic`](https://crates.io/crates/forward_ref_generic). This is used [in the compiler itself](https://github.com/rust-lang/rust/blob/4f7bb9890c0402cd145556ac1929d13d7524959e/library/core/src/internal_macros.rs#L4)